### PR TITLE
Improve offline editing layer detection and code

### PR DESCRIPTION
**Pull request by m-kuhn**
Offline editing currently relies on

- "layername" being the exactly the first part of the URL
- `QgsDataSourceUri::param( "{file_path_to_gpkg}|layername" )` to return the layer name

This solidifies this behavior with the regex approach from the gpkg provider (still not perfect but *much* more stable).

Also comes with some general modernization of the code.

### CODE REVIEW

File: src/core/qgsofflineediting.cpp [hunk 1]
```diff
@@ -1492,7 +1492,15 @@ void QgsOfflineEditing::committedFeaturesAdded( const QString &qgisLayerId, cons
   }
   else
   {
-    tableName = uri.param( offlinePath + "|layername" );
+    if ( dataSourceString.indexOf( QLatin1String( "|layername=" ) ) != -1 )
```
Comment (nyalldawson) [inline src/core/qgsofflineediting.cpp]: Why not use QgsDataProviderRegistry::decodeUri instead?

Comment (m-kuhn) [reply src/core/qgsofflineediting.cpp]: I was looking for something like `layer->dataProvider().staticMetaData.decodeUri()` but didn't find any.

Comment (nyalldawson) [reply src/core/qgsofflineediting.cpp]: Nope, you do it through the registry using the layer's provider key

Comment (m-kuhn) [reply src/core/qgsofflineediting.cpp]: Ok. If this is the preferred approach I can do that.

Comment (nyalldawson) [reply src/core/qgsofflineediting.cpp]: Yep, it's definitely preferred. If nothing else it's covered by unit tests...

**Review by signedav — APPROVED** [review]

**Comment by m-kuhn** [comment]
Manually tested and works

File: src/core/qgsofflineediting.cpp (hunk 1 above)
Comment (m-kuhn) [reply src/core/qgsofflineediting.cpp]: Done :+1:

**Pull request merged**